### ISSUE TITLE
Compile issue with GCC 11.2.0 in C++17 projects

### ISSUE THREAD

**Issue by 1uc** [BlueBrain/libsonata#246]
When using libsonata in C++17 projects, those projects fail to compile due to `libsonata`. A minimal reproducer for this issue is:
```
// filename: compile_issue.cpp
#include <bbp/sonata/config.h>

int main() {
  bbp::sonata::CircuitConfig::fromFile("foo.json");
  return 0;
}
```

One can obtain an affected version of libsonata via Spack:
```
spack install --jobs $(nproc) libsonata
```

and compile the reproducer as follows:
```
g++ --std=c++17 -I $(spack location --install-dir libsonata)/include -L $(spack location --install-dir libsonata)/lib compile-issue.cpp -lsonata -o compile-issue
```

The output is:
```
# ...
/gpfs/bbp.cscs.ch/project/proj16/scratch/groshein/spack_install/software/install_gcc-11.2.0-skylake/libsonata-develop-k5b3q3/include/bbp/sonata/config.h:715:14:   required from here
/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/type_traits:906:30: error: default member initializer for 'bbp::sonata::SimulationConfig::InputBase::delay' required before the end of its enclosing class
  906 |     : public __bool_constant<__is_constructible(_Tp, _Args...)>
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from main.cpp:1:
/gpfs/bbp.cscs.ch/project/proj16/scratch/groshein/spack_install/software/install_gcc-11.2.0-skylake/libsonata-develop-k5b3q3/include/bbp/sonata/config.h:375:21: note: defined here
  375 |         double delay{};
      |                     ^~~
In file included from /gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/move.h:57,
                 from /gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_pair.h:59,
                 from /gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/bits/stl_algobase.h:64,
                 from /gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/memory:63,
                 from /gpfs/bbp.cscs.ch/project/proj16/scratch/groshein/spack_install/software/install_gcc-11.2.0-skylake/libsonata-develop-k5b3q3/include/bbp/sonata/config.h:14,
                 from main.cpp:1:
/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_compilers/install_gcc-4.8.5-haswell/gcc-11.2.0-ehzq5x/lib/gcc/x86_64-pc-linux-gnu/11.2.0/../../../../include/c++/11.2.0/type_traits:906:30: error: default member initializer for 'bbp::sonata::SimulationConfig::InputBase::duration' required before the end of its enclosing class
  906 |     : public __bool_constant<__is_constructible(_Tp, _Args...)>
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from main.cpp:1:
/gpfs/bbp.cscs.ch/project/proj16/scratch/groshein/spack_install/software/install_gcc-11.2.0-skylake/libsonata-develop-k5b3q3/include/bbp/sonata/config.h:377:24: note: defined here
  377 |         double duration{};
      |                        ^~~
```

The issue seems to be the following construct:

```
class SONATA_API SimulationConfig
{
  public:
      struct InputBase {
        /// Time when input is activated (ms)
        double delay{};
        /// Time duration for how long input is activated (ms)
        double duration{};
    };

    struct InputLinear: public InputBase {
       // ...
    };
```

We can find a similar issue reported here:
https://stackoverflow.com/questions/53408962/try-to-understand-compiler-error-message-default-member-initializer-required-be

However, the compiler seems to not complain about earlier cases where members of a member structs are default/value initialized. Only for `InputBase` which is later inherited by other member structs. I've tried applying the proposed fix of moving the member struct to a "protected" namespace. We seem to only need to do this for `InputBase`. Roughly,

```
namespace detail {
struct InputBase {
  /// Time when input is activated (ms)
  double delay{};
  /// Time duration for how long input is activated (ms)
  double duration{};
};
}

class SONATA_API SimulationConfig
{
  public:
    using InputBase = detail::InputBase;

    struct InputLinear: public InputBase {
       // ...
    };
```

This will enable the reproducer to compile. I'm happy to provide a PR if this type of change is deems acceptable.

**Comment by 1uc** [comment]
This is likely a compiler issue in GCC 11.2.0 that has been fixed in GCC 12.2.0. Since the latter is currently being deployed at BBP, no changes are required.

**Comment by mgeplf** [comment]
Thanks, must have been annoying to track this down.
